### PR TITLE
FEATURE: Allow selective dismissal of new and unread topics

### PR DESCRIPTION
**Pull request by martin-brennan**
This PR improves the UI of bulk select so that its context is applied to the Dismiss Unread and Dismiss New buttons. Regular users (not just staff) are now able to use topic bulk selection on the /new and /unread routes to perform these dismiss actions more selectively.

For Dismiss Unread, there is a new count in the text of the button and in the modal when one or more topic is selected with the bulk select checkboxes.

image
image

For Dismiss New, there is a count in the button text, and we have added functionality to the server side to accept an array of topic ids to dismiss new for, instead of always having to dismiss all new, the same as the bulk dismiss unread functionality. To clean things up, the `DismissTopics` service has been rolled into the `TopicsBulkAction` service.

We now also show the top Dismiss/Dismiss New button based on whether the bottom one is in the viewport, not just based on the topic count.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/lib/is-element-in-viewport.js [hunk 1]
```diff
@@ -1,4 +1,7 @@
 export default function (element) {
+  if (!element) {
+    return;
+  }
```
Comment (SamSaffron) [inline app/assets/javascripts/discourse/app/lib/is-element-in-viewport.js:4]: unrelated: I wonder if we should clean up usages of this API so we never call it with jQuery objects, seems like a superfluous test, just send it the right type of object.

File: app/controllers/topics_controller.rb [hunk 2]
```diff
@@ -970,7 +970,12 @@ def reset_new
         end
       end
 
-    dismissed_topic_ids = DismissTopics.new(current_user, topic_scope).perform!
+    if params[:topic_ids].present?
+      topic_ids = params[:topic_ids].map { |t| t.to_i }
```
Comment (SamSaffron) [inline app/controllers/topics_controller.rb:974]: will this end up with a 500 error if you do `topic_ids=1` ? do we want to ensure it is an array?

File: lib/topics_bulk_action.rb [hunk 3]
```diff
@@ -81,6 +81,24 @@ def dismiss_posts
...
+      .where("topics.archetype <> ?", Archetype.private_message)
+      .order("topics.created_at DESC")
+      .limit(SiteSetting.max_new_topics).map do |topic|
+      {
+        topic_id: topic.id,
+        user_id: @user.id,
+        created_at: Time.zone.now
+      }
+    end
+    DismissedTopicUser.insert_all(rows) if rows.present?
```
Comment (SamSaffron) [inline lib/topics_bulk_action.rb:98]: do we want to allow for duplicates and race conditions here?

**Comment by SamSaffron** [comment]
server side looks good minus a few minor comments, @CvX can you review the client side?

File: app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        );
+        this.set(
+          "isOtherDismissNewButtonVisible",
+          isElementInViewport(document.getElementById("dismiss-new-bottom"))
+        );
+      } else {
+        this.set("isOtherDismissUnreadButtonVisible", true);
+        this.set("isOtherDismissNewButtonVisible", true);
+      }
+    });
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js:89]:
```suggestion
  @afterRender
  _determineOtherDismissVisibility() {
    if (this.position === "top") {
      this.set(
        "isOtherDismissUnreadButtonVisible",
        isElementInViewport(document.getElementById("dismiss-topics-bottom"))
      );
      this.set(
        "isOtherDismissNewButtonVisible",
        isElementInViewport(document.getElementById("dismiss-new-bottom"))
      );
    } else {
      this.set("isOtherDismissUnreadButtonVisible", true);
      this.set("isOtherDismissNewButtonVisible", true);
    }
```

(needs: `import { afterRender } from "discourse-common/utils/decorators";`)

File: app/assets/javascripts/discourse/app/mixins/bulk-topic-selection.js [hunk 5]
```diff
@@ -12,13 +11,23 @@ export default Mixin.create({
   autoAddTopicsToBulkSelect: false,
   selected: null,
 
-  canBulkSelect: alias("currentUser.staff"),
+  @discourseComputed("currentUser.staff", "showDismissRead", "showResetNew")
+  canBulkSelect(isStaff, showDismissRead, showResetNew) {
+    return isStaff || showDismissRead || showResetNew;
+  },
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/mixins/bulk-topic-selection.js:17]: Looks like a job for `or()`

File: app/assets/javascripts/discourse/app/mixins/bulk-topic-selection.js [hunk 6]
```diff
@@ -12,13 +11,23 @@ export default Mixin.create({
...
   @on("init")
   resetSelected() {
     this.set("selected", []);
   },
 
+  _isFilterPage(filter, filterType) {
+    if (!filter) {
+      return false;
+    }
+    return filter.match(new RegExp(filterType + "$", "gi")) ? true : false;
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/mixins/bulk-topic-selection.js:28]:
```suggestion
    return new RegExp(filterType + "$", "gi").test(filter);
```

File: app/assets/javascripts/discourse/app/models/topic.js [hunk 7]
```diff
@@ -756,7 +756,15 @@ Topic.reopenClass({
...
 
-  resetNew(category, include_subcategories, tracked = false, tag = false) {
+  resetNew(category, include_subcategories, opts = {}) {
+    let tracked, tag, topicIds;
+
+    // apply default options
+    ({ tracked, tag, topicIds } = Object.assign(
+      { tracked: false, tag: null, topicIds: null },
+      opts
+    ));
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/models/topic.js:766]: Would this work?

```suggestion
    let { tracked, tag, topicIds } = {
      tracked: false,
      tag: null,
      topicIds: null,
      ...opts
    };
```

**Review by CvX — CHANGES_REQUESTED** [review]

File: app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js (hunk 4 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js:90]: @CvX afterRender did not work for me at all here for some reason, the code inside the afterRender function was never executed?

Comment (CvX) [reply app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js:90]: 🤦 Gah, my bad! I always forget how `@afterRender` works! It's not "run this code after rendering" but "when called, delay the execution after rendering". So the correct suggestion would be either:

```js
  didInsertElement() {
    _super.didInsertElement(...arguments);

    // code
```

or:

```js
  init() {
    this._super(...arguments);
    this._determineOtherDismissVisibility();
  }

  @afterRender
  _determineOtherDismissVisibility() {
    // code
```

File: app/assets/javascripts/discourse/app/lib/is-element-in-viewport.js (hunk 1 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/lib/is-element-in-viewport.js:4]: Agreed, removing this

File: lib/topics_bulk_action.rb (hunk 3 above)
Comment (martin-brennan) [reply lib/topics_bulk_action.rb:98]: @SamSaffron based on the docs https://apidock.com/rails/v6.0.0/ActiveRecord/Persistence/ClassMethods/insert_all and other usages of insert_all in our code, it skips duplicates by default, so I don't think there is anything we need to do here.

https://github.com/discourse/discourse/blob/a1ea4a3b4ef0862c69207f4a08661ffea66c34c8/app/models/topic_thumbnail.rb#L24-L30

File: app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js (hunk 4 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js:90]: Cool, will just leave it as is then :)

**Comment by martin-brennan** [comment]
@SamSaffron and @CvX can you both please take another look at this, I have addressed the review comments.

File: app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js (hunk 4 above)
Comment (CvX) [reply app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js:90]: Is the `later()` call required though?

Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/components/topic-dismiss-buttons.js:90]: As far as I could tell, yes. If I removed it I did not get the desired result — the other element is not counted as visible even though it clearly is.

**Review by CvX — APPROVED** [review]

**Pull request merged**